File: lang/c/src/vtfppc/PSFrames.cpp
#include <vtfppc/PSFrames.h>

#include <sourceppc/Helpers.h>

using namespace sourceppc;
using namespace vtfpp;

SOURCEPP_API vtfpp_psframes_handle_t vtfpp_psframes_open_from_mem(const unsigned char* buffer, size_t bufferLen) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, nullptr);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, nullptr);

	return new PSFrames{{reinterpret_cast<const std::byte*>(buffer), bufferLen}};
}

SOURCEPP_API vtfpp_psframes_handle_t vtfpp_psframes_open_from_file(const char* psframesPath) {
	SOURCEPP_EARLY_RETURN_VAL(psframesPath, nullptr);

	return new PSFrames{psframesPath};
}

SOURCEPP_API void vtfpp_psframes_close(vtfpp_psframes_handle_t* handle) {
	SOURCEPP_EARLY_RETURN(handle);

	delete convert::handle<PSFrames>(*handle);
	*handle = nullptr;
}

SOURCEPP_API uint32_t vtfpp_psframes_get_frame_count(vtfpp_psframes_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	return convert::handle<PSFrames>(handle)->getFrameCount();
}

SOURCEPP_API uint32_t vtfpp_psframes_get_fps(vtfpp_psframes_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	return convert::handle<PSFrames>(handle)->getFPS();
}

SOURCEPP_API uint16_t vtfpp_psframes_get_width(vtfpp_psframes_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	return convert::handle<PSFrames>(handle)->getWidth();
}

SOURCEPP_API uint16_t vtfpp_psframes_get_height(vtfpp_psframes_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	return convert::handle<PSFrames>(handle)->getHeight();
}

SOURCEPP_API sourcepp_buffer_t vtfpp_psframes_get_image_data_as(vtfpp_psframes_handle_t handle, vtfpp_image_format_e newFormat, uint32_t frame) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_BUFFER_INVALID);

	return convert::toBuffer(convert::handle<PSFrames>(handle)->getImageDataAs(convert::cast(newFormat), frame));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_psframes_get_image_data_as_bgr888(vtfpp_psframes_handle_t handle, uint32_t frame) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_BUFFER_INVALID);

	return convert::toBuffer(convert::handle<PSFrames>(handle)->getImageDataAsBGR888(frame));
}

File: lang/csharp/src/vtfpp/PSFrames.cs
// ReSharper disable InconsistentNaming

using System;

namespace sourcepp.vtfpp;

public sealed class PSFrames : sourcepp.ManagedNativeHandle
{
	internal PSFrames(nint handle) : base(handle, DLL.vtfpp_psframes_close)
	{
	}

	public PSFrames(ReadOnlySpan<byte> buffer) : this(DLL.vtfpp_psframes_open_from_mem(buffer, (ulong) buffer.Length))
	{
	}

	public PSFrames(string psframesPath) : this(DLL.vtfpp_psframes_open_from_file(psframesPath))
	{
	}

	public static implicit operator bool(PSFrames frames)
	{
		frames.ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_psframes_is_valid(frames.Handle));
	}

	public uint FrameCount
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vtfpp_psframes_get_frame_count(Handle);
		}
	}

	public uint FPS
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vtfpp_psframes_get_fps(Handle);
		}
	}

	public ushort GetWidth(uint frame)
	{
		ThrowIfDisposed();
		return DLL.vtfpp_psframes_get_width(Handle, frame);
	}

	public ushort GetHeight(uint frame)
	{
		ThrowIfDisposed();
		return DLL.vtfpp_psframes_get_height(Handle, frame);
	}

	public byte[] GetPaletteDataRaw(uint frame)
	{
		ThrowIfDisposed();
		return new sourcepp.Buffer(DLL.vtfpp_psframes_get_palette_data_raw(Handle, frame)).Read<byte>();
	}

	public byte[] GetPaletteDataAs(ImageFormat newFormat, uint frame)
	{
		ThrowIfDisposed();
		return new sourcepp.Buffer(DLL.vtfpp_psframes_get_palette_data_as(Handle, newFormat, frame)).Read<byte>();
	}

	public byte[] GetImageDataRaw(uint frame)
	{
		ThrowIfDisposed();
		return new sourcepp.Buffer(DLL.vtfpp_psframes_get_image_data_raw(Handle, frame)).Read<byte>();
	}

	public byte[] GetImageDataAs(ImageFormat newFormat, uint frame)
	{
		ThrowIfDisposed();
		return new sourcepp.Buffer(DLL.vtfpp_psframes_get_image_data_as(Handle, newFormat, frame)).Read<byte>();
	}

	public byte[] GetImageDataAsBGR888(uint frame)
	{
		ThrowIfDisposed();
		return new sourcepp.Buffer(DLL.vtfpp_psframes_get_image_data_as_bgr888(Handle, frame)).Read<byte>();
	}
}
